bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/StateDescriptionServiceImpl.java

```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.eclipse.smarthome.core.service;
+
+import java.math.BigDecimal;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.TreeSet;
+
+import org.eclipse.smarthome.core.types.StateDescription;
+import org.eclipse.smarthome.core.types.StateDescriptionProvider;
+import org.eclipse.smarthome.core.types.StateOption;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+
+/**
+ * @author Lyubomir Papazov - Initial contribution
+ *
+ */
+@Component()
+public class StateDescriptionServiceImpl implements StateDescriptionService {
+
+    private final Set<StateDescriptionProvider> stateDescriptionProviders = Collections
+            .synchronizedSet(new TreeSet<StateDescriptionProvider>(new Comparator<StateDescriptionProvider>() {
+        @Override
+        public int compare(StateDescriptionProvider provider1, StateDescriptionProvider provider2) {
+            return provider2.getRank().compareTo(provider1.getRank());
+        }
+    }));
+        
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    public void addStateDescriptionProvider(StateDescriptionProvider provider) {
+        stateDescriptionProviders.add(provider);
+    }
+
+    public void removeStateDescriptionProvider(StateDescriptionProvider provider) {
+        stateDescriptionProviders.remove(provider);
+    }
+
+    @Override
+    public StateDescription getStateDescription(String itemName, Locale locale) {
+        BigDecimal minimum = null;
+        BigDecimal maximum = null;
+        BigDecimal step = null;
+        String pattern = null;
+        boolean isReadOnly = false;
+        List<StateOption> stateOptions = Collections.emptyList();
+
+        boolean foundNonNullStateDescription = false;
+
+        for (StateDescriptionProvider stateDescriptionProvider : stateDescriptionProviders) {
+            StateDescription currentDescription = stateDescriptionProvider.getStateDescription(itemName, locale);
+            if (currentDescription != null) {
+                foundNonNullStateDescription = true;
+                if (shouldCheckNextDescriptionProvider(minimum, maximum, step, pattern, isReadOnly, stateOptions)) {
```

Comment:
matter of taste, but i would like this to be turned around, like `if(allPropertiesSet(...)) break;` and otherwise continue with the overwrite logic.

Reply:
Besides that: the new merge logic might result in different behaviour and makes this API breaking I guess. wdyt?